### ISSUE TITLE
Examples/descriptions for typescript 3.6.3

### ISSUE THREAD

**Issue by simplone** [redux-saga/redux-saga#1932]
### Description
I have used only few weeks redux saga and tried to update packages for latest versions. But stuck with upgrade of typescript version. In latest added update for typechecks in generators. 
https://devblogs.microsoft.com/typescript/announcing-typescript-3-6/#stricter-generators
Please provide, if it possible, more info how to use current d.ts files and match types. I hope anybody from community had the same troubles. 

Thanks

**Comment by Andarist** [comment]
Some examples can be found here - https://github.com/redux-saga/redux-saga/pull/1892#issue-297396449

**Comment by kristian-puccio** [comment]
Is there an example of how to get the select effect to propagate state?
Without any extra type annotations I'm getting an any type.

To work around this I'm currently using ReturnType<typeof mySelectorFunc> to annotate the resulting value.

Thanks!

**Comment by Andarist** [comment]
@kristian-puccio this is not yet possible with TS - it can't express a relation between yielded value and the injected one.

**Comment by kristian-puccio** [comment]
Thanks @Andarist for the clarification. 
I'll go look up the TS issues to see if it's on the roadmap

**Comment by henrikra** [comment]
@Andarist But looks like there is this minimal wrapper which makes it possible? https://github.com/agiledigital/typed-redux-saga

**Comment by Andarist** [comment]
Yes, this is worth exploring. I would like to get input on this idea from more people first though. cc @aikoven @gilbsgilbs

**Comment by kristian-puccio** [comment]
@henrikra Awesome! I'm currently deleting all my type definitions from inside my sagas. Soo good!

I have no idea what the `yield*` is about though and I don't like how if you forget the * it defaults to any, I would prefer unknown.

**Comment by gilbsgilbs** [comment]
@kristian-puccio the [`yield*` operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/yield*) is similar to `yield`, except it yields all the values from an iterable instead of one value.

e.g.

```js
function* gen0() {
  yield 1;
  yield 2;
}

// is equivalent to

function* gen1() {
  yield* [1, 2];
}
```

It happens that the yield delegate operator has a slightly different (and simpler) behavior than the yield operator regarding return types. With the `yield` operator, the return type depends on what argument is passed to the `.next(…)` method which is very volatile (and unsupported by TypeScript at this date as you noticed). With `yield*`, [the return type is the return type of the invoked generator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/yield*#The_value_of_yield*_expression_itself). Thanks to this nice property, TypeScript 3.6 brings complete support for strong types on the `yield*` operator. `typed-redux-saga` relies on this and wraps redux-saga's effects to strongly type their return types.

>  I don't like how if you forget the * it defaults to any

This cannot be avoided I think. We need linting rules. If typed-redux-saga exported every effect from `redux-saga/effects`, we could actually use the [`no-restricted-imports`](https://eslint.org/docs/rules/no-restricted-imports) rule from ESLint. It would also reduce mental charge (“do I have to use `yield` or `yield*` for this effect?”).

@Andarist I think we need to keep this approach a bit "underground" for the time being, just to play around and see how well it fits to various use-cases. I'm quite optimistic. I don't think strong types for `yield` return types are coming into TS anytime soon, so I think it's the best approach. When it'll reach a higher level of stability, maybe we could consider migrating it into `redux-saga` workspace (if @danielnixon feels comfortable with this) and documenting it in redux-saga documentation directly.

**Comment by kristian-puccio** [comment]
@gilbsgilbs Thank you so much for taking the time to write the detailed explanation.

**Comment by danielnixon** [comment]
> If typed-redux-saga exported every effect from redux-saga/effects
PRs welcome ;-)
> When it'll reach a higher level of stability, maybe we could consider migrating it into redux-saga workspace (if @danielnixon feels comfortable with this)
Yep, definitely :)

**Comment by danielnixon** [comment]
> I don't like how if you forget the * it defaults to any, I would prefer unknown.

> This cannot be avoided I think. We need linting rules.

@kristian-puccio @gilbsgilbs take a look at:

* https://github.com/plantain-00/type-coverage
* https://palantir.github.io/tslint/rules/no-any/ and https://palantir.github.io/tslint/rules/no-unsafe-any/ (if you still use tslint)
* https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-explicit-any.md (if you've migrated to eslint)

I use these to great effect to ban `any`s.

**Comment by henrikra** [comment]
> > If typed-redux-saga exported every effect from redux-saga/effects
> 
> PRs welcome ;-)
> 
> > When it'll reach a higher level of stability, maybe we could consider migrating it into redux-saga workspace (if @danielnixon feels comfortable with this)
> 
> Yep, definitely. `typed-redux-saga` was always a science experiment and I'm happy to retire it if/when the same capability makes its way into redux-saga natively. :)

What redux-saga should do in order to retire the `typed-redux-saga`?

**Comment by danielnixon** [comment]
> What redux-saga should do in order to retire the typed-redux-saga?
Expose `yield*` (generator) versions of all the built in redux-saga effects so we can take advantage of the improved type safety that @gilbsgilbs describes in https://github.com/redux-saga/redux-saga/issues/1932#issuecomment-533096118

**Comment by henrikra** [comment]
Okay so if you know how to do it maybe create PR for Redux-saga and tell write in the description this use case so we could get it merged? 🤩

**Comment by donifer** [comment]
Is the idea to still keep `typed-redux-saga` method "underground"? If so, is there a plan moving forward to fully recommend or implement directly in `redux-saga`?

**Comment by henrikra** [comment]
@donifer This is exactly what I would like to know ☝️

**Comment by Andarist** [comment]
@gilbsgilbs @aikoven WDYT about this idea? Seems reasonable to me - but I'd like to have some kind of confirmation from you guys (you are way more experienced with TS than I am)

**Comment by aikoven** [comment]
I like the idea. I don’t expect that TS would get support for strongly typing yields. This solution is a bit of a hack, but it involves very little code from user side. It would be nice to have official support for it. But I think that we should keep this as a separate package to not confuse JS users.

**Comment by henrikra** [comment]
Sure and there has to be some paragraph in redux-saga docs pointing TS users to this separate repo for better type support imo 👍

**Comment by Andarist** [comment]
I’m inclined into incorporating this into this monorepo (if @danielnixon allows for that) - just this would have to be a separate package, smth like @redux-saga/typed-core. We could also support rewriting yield* typedEffect to regular yield effect with our babel plugin - to save some bytes and runtime indirection. The typed package is neccessary only to satisfy TS (it can and should work without compile-time rewrites though)

**Comment by donifer** [comment]
I love the idea of the Babel plugin.

**Comment by Andarist** [comment]
Keep in mind that babel plugin won't enhance the experience around this dramatically - it will only optimize the output **a little bit**.

**Comment by TeoTN** [comment]
I mention it also [here](https://github.com/redux-saga/redux-saga/issues/1976#issuecomment-561654494) but I think TS doesn't allow you to type every `yield` expression in a generator separately.

Even if it allowed that, it'd be probably up to the users to declare types of subsequent effects calls, since current implementation (added in TS 3.6) allows specifying a single type for all yield values, but it doesn't mean that middleware would infer the type returned from `select` or `call` and "pass" it back to the variable declaration in saga, that the yield is assigned to.

What I think is possible though is changing the API drastically so that effects return generators:

```
function* newCall<Fn extends (...args: any[]) => any>(
    fn: Fn,
    ...args: Parameters<Fn>
): Generator<Effect, ReturnType<Fn>, ReturnType<Fn>> {
    return yield call<Fn>(fn, ...args);
}

function* saga() {
  const result = yield* newCall(() => 5); // `result` type is inferred to be number
}
```

**Comment by pooyahrtn** [comment]
> I mention it also [here](https://github.com/redux-saga/redux-saga/issues/1976#issuecomment-561654494) but I think TS doesn't allow you to type every `yield` expression in a generator separately.
> 
> Even if it allowed that, it'd be probably up to the users to declare types of subsequent effects calls, since current implementation (added in TS 3.6) allows specifying a single type for all yield values, but it doesn't mean that middleware would infer the type returned from `select` or `call` and "pass" it back to the variable declaration in saga, that the yield is assigned to.
> 
> What I think is possible though is changing the API drastically so that effects become generators:
> 
> ```
> function* newCall<Fn extends (...args: any[]) => any>(
>     fn: Fn,
>     ...args: Parameters<Fn>
> ): Generator<Effect, ReturnType<Fn>, ReturnType<Fn>> {
>     return yield call<Fn>(fn, ...args);
> }
> 
> function* saga() {
>   const result = yield* newCall(() => 5); // `result` type is inferred to be number
> }
> ```

Hi, Im new to redux-saga and I'm not sure about this, but since call results could be promise, I made a few changes to your code that might help someone.
```typescript
type Unpromisify<T> = T extends Promise<infer K> ? K : T;
type SagaCallReturn<T extends (...args: any) => any> = Unpromisify<
  ReturnType<T>
>;
export function* callStar<Fn extends (...args: any[]) => any>(
  fn: Fn,
  ...args: Parameters<Fn>
): Generator<Effect, SagaCallReturn<Fn>, SagaCallReturn<Fn>> {
  return yield call<Fn>(fn, ...args);
}
```
